File: Python/Tree/M-199. Binary Tree Right Side View.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []

        values: List[int] = []

        def bfs(nodes: List[TreeNode]) -> None:
            if len(nodes) == 0:
                return

            nonlocal values
            values.append(nodes[len(nodes) - 1].val)

            next_nodes: List[TreeNode] = []
            while len(nodes):
                node = nodes.pop(0)
                if node.left:
                    next_nodes.append(node.left)
                if node.right:
                    next_nodes.append(node.right)
            bfs(next_nodes)

        bfs([root])
        return values

File: Javascript/Tree/M-199. Binary Tree Right Side View.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

//      1
//  2       3
//    4   8   6
//  2   5

function rightSideView(root: TreeNode | null): number[] {
  if (!root) return [];

  const values: number[] = [];
  const bfs = (nodes: TreeNode[]): void => {
    if (!nodes.length) return;

    /** push the rightest node value */
    values.push(nodes[nodes.length - 1].val);

    /** prepare the nodes of next level  */
    const nextNodes: TreeNode[] = [];
    while (nodes.length) {
      const node = nodes.shift();
      if (node?.left) nextNodes.push(node.left);
      if (node?.right) nextNodes.push(node.right);
    }

    /** travelsal to next level */
    bfs(nextNodes);
  };

  bfs([root]);
  return values;
}
